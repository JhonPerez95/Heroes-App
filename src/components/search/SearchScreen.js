import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import PropTypes from 'prop-types';

import HeroeCard from '../heroes/HeroeCard';
import useForm from '../../hook/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';

const SearchScreen = ({ history }) => {
  // Location params
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);

  // Form
  const [formValue, handleInputChange] = useForm({
    findHero: q,
  });
  const { findHero } = formValue;

  // Selector
  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  const hanldeSubmit = (e) => {
    e.preventDefault();
    history.push(`?q=${findHero}`);
  };

  return (
    <div>
      <h1> Search Screen: </h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4> Search Form</h4>
          <form onSubmit={hanldeSubmit}>
            <input
              name="findHero"
              type="text"
              placeholder="Find your hero"
              className="form-control"
              value={findHero}
              onChange={handleInputChange}
              autoComplete="off"
            />
            <button type="submit" className="btn btn-block btn-outline-primary">
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {q === '' && <div className="alert alert-info">Search a hero !</div>}
          {q !== '' && heroesFiltered.length === 0 && (
            <div className="alert alert-danger">There is no hero with {q}</div>
          )}
          {heroesFiltered.map((heroe) => (
            <HeroeCard key={heroe.id} {...heroe} />
          ))}
        </div>
      </div>
    </div>
  );
};

SearchScreen.propTypes = {
  history: PropTypes.object.isRequired,
};

export default SearchScreen;
