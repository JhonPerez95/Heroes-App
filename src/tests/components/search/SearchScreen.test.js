import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';

import SearchScreen from '../../../components/search/SearchScreen';

describe('Test a component <SearchScreen/>', () => {
  test('should correctly the component', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero !');
  });

  test('should show to batman and a input with valor of the query ', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper).toMatchSnapshot();
  });

  test('should show a error if not found a hero', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman12132']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find('.alert-danger').exists()).toBe(true);
    expect(wrapper.find('.alert-info').exists()).toBe(false);
  });

  test('should called the push history', () => {
    const history = {
      push: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route
          path="/search"
          component={() => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
      target: {
        name: 'findHero',
        value: 'batman',
      },
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault() {},
    });

    expect(history.push).toHaveBeenCalledWith('?q=batman');
  });
});
