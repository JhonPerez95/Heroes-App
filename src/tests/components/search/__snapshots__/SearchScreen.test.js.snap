// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test a component <SearchScreen/> should correctly the component 1`] = `
<div>
  <h1>
     Search Screen: 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="findHero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero !
      </div>
    </div>
  </div>
</div>
`;

exports[`Test a component <SearchScreen/> should show to batman and a input with valor of the query  1`] = `
<div>
  <h1>
     Search Screen: 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="findHero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img-top"
              src="../assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/heroe/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
